In [1]:
#import necessary library
import  numpy as np
import pandas as pd
import matplotlib as plt
import seaborn as sns

In [2]:
df = pd.read_csv("movie_metadata.csv")

In [3]:
df.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating        budget  \
0               3054.0  English      USA           PG-13  237000000.0$   
1               1238.0  English      USA           PG-13  300000000.0$   
2                994.0  English       UK           PG-13  245000000.0$   
3               2701.0  English      USA           PG-13  250000000.0$   
4                  NaN      NaN      NaN             NaN          0.0$   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [4]:
df.tail()

color     director_name  num_critic_for_reviews  duration  \
5038  Color       Scott Smith                     1.0      87.0   
5039  Color               NaN                    43.0      43.0   
5040  Color  Benjamin Roberds                    13.0      76.0   
5041  Color       Daniel Hsia                    14.0     100.0   
5042  Color          Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
5038                      2.0                   318.0     Daphne Zuniga   
5039                      NaN                   319.0     Valorie Curry   
5040                      0.0                     0.0     Maxwell Moody   
5041                      0.0                   489.0     Daniel Henney   
5042                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes    gross                        genres  ...  \
5038                   637.0      NaN                  Comedy|Drama  ...   
5039                   841.0      NaN  Crime|Drama|Mystery|Thriller  ...   
5040                     0.0      NaN         Drama|Horror|Thriller  ...   
5041                   946.0  10443.0          Comedy|Drama|Romance  ...   
5042                    86.0  85222.0                   Documentary  ...   

     num_user_for_reviews language  country  content_rating   budget  \
5038                  6.0  English   Canada             NaN     0.0$   
5039                359.0  English      USA           TV-14     0.0$   
5040                  3.0  English      USA             NaN  1400.0$   
5041                  9.0  English      USA           PG-13     0.0$   
5042                 84.0  English      USA              PG  1100.0$   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
5038      2013.0                  470.0        7.7           NaN   
5039         NaN                  593.0        7.5         16.00   
5040      2013.0                    0.0        6.3           NaN   
5041      2012.0                  719.0        6.3          2.35   
5042      2004.0                   23.0        6.6          1.85   

     movie_facebook_likes  
5038                   84  
5039                32000  
5040                   16  
5041                  660  
5042                  456  

[5 rows x 28 columns]

In [5]:
df.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [6]:
df.describe()

num_critic_for_reviews     duration  director_facebook_likes  \
count             4993.000000  5028.000000              4939.000000   
mean               140.194272   107.201074               686.509212   
std                121.601675    25.197441              2813.328607   
min                  1.000000     7.000000                 0.000000   
25%                 50.000000    93.000000                 7.000000   
50%                110.000000   103.000000                49.000000   
75%                195.000000   118.000000               194.500000   
max                813.000000   511.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             5020.000000             5036.000000  4.159000e+03   
mean               645.009761             6560.047061  4.846841e+07   
std               1665.041728            15020.759120  6.845299e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                133.000000              614.000000  5.340988e+06   
50%                371.500000              988.000000  2.551750e+07   
75%                636.000000            11000.000000  6.230944e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     5.043000e+03                5043.000000           5030.000000   
mean      8.366816e+04                9699.063851              1.371173   
std       1.384853e+05               18163.799124              2.013576   
min       5.000000e+00                   0.000000              0.000000   
25%       8.593500e+03                1411.000000              0.000000   
50%       3.435900e+04                3090.000000              1.000000   
75%       9.630900e+04               13756.500000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews   title_year  actor_2_facebook_likes   imdb_score  \
count           5022.000000  4935.000000             5030.000000  5043.000000   
mean             272.770808  2002.470517             1651.754473     6.442138   
std              377.982886    12.474599             4042.438863     1.125116   
min                1.000000  1916.000000                0.000000     1.600000   
25%               65.000000  1999.000000              281.000000     5.800000   
50%              156.000000  2005.000000              595.000000     6.600000   
75%              326.000000  2011.000000              918.000000     7.200000   
max             5060.000000  2016.000000           137000.000000     9.500000   

       aspect_ratio  movie_facebook_likes  
count   4714.000000           5043.000000  
mean       2.220403           7525.964505  
std        1.385113          19320.445110  
min        1.180000              0.000000  
25%        1.850000              0.000000  
50%        2.350000            166.000000  
75%        2.350000           3000.000000  
max       16.000000         349000.000000

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      5024 non-null   object 
 1   director_name              4939 non-null   object 
 2   num_critic_for_reviews     4993 non-null   float64
 3   duration                   5028 non-null   float64
 4   director_facebook_likes    4939 non-null   float64
 5   actor_3_facebook_likes     5020 non-null   float64
 6   actor_2_name               5030 non-null   object 
 7   actor_1_facebook_likes     5036 non-null   float64
 8   gross                      4159 non-null   float64
 9   genres                     5043 non-null   object 
 10  actor_1_name               5036 non-null   object 
 11  movie_title                5043 non-null   object 
 12  num_voted_users            5043 non-null   int64  
 13  cast_total_facebook_likes  5043 non-null   int64

In [8]:
df.shape

(5043, 28)

In [9]:
df.dtypes

color                         object
director_name                 object
num_critic_for_reviews       float64
duration                     float64
director_facebook_likes      float64
actor_3_facebook_likes       float64
actor_2_name                  object
actor_1_facebook_likes       float64
gross                        float64
genres                        object
actor_1_name                  object
movie_title                   object
num_voted_users                int64
cast_total_facebook_likes      int64
actor_3_name                  object
facenumber_in_poster         float64
plot_keywords                 object
movie_imdb_link               object
num_user_for_reviews         float64
language                      object
country                       object
content_rating                object
budget                        object
title_year                   float64
actor_2_facebook_likes       float64
imdb_score                   float64
aspect_ratio                 float64
m

In [10]:
df.isnull().sum()

color                         19
director_name                104
num_critic_for_reviews        50
duration                      15
director_facebook_likes      104
actor_3_facebook_likes        23
actor_2_name                  13
actor_1_facebook_likes         7
gross                        884
genres                         0
actor_1_name                   7
movie_title                    0
num_voted_users                0
cast_total_facebook_likes      0
actor_3_name                  23
facenumber_in_poster          13
plot_keywords                153
movie_imdb_link                0
num_user_for_reviews          21
language                      12
country                        5
content_rating               303
budget                         0
title_year                   108
actor_2_facebook_likes        13
imdb_score                     0
aspect_ratio                 329
movie_facebook_likes           0
dtype: int64

1.2. The duration column contains data on the film length. How many missing values are there in this column?

How many missing values are there in Duration?

In [11]:
df["duration"].isnull().sum()

15

There are fifteen missing values in duration

1.3. Replace the missing values in the duration column with the median value for this column.

Replace the missing values in the duration column with the median value for this column.

In [12]:
df["duration"] = df["duration"].fillna(df["duration"].median())

In [13]:
df["duration"].isnull().sum()

0

1.4. What is the average film length? Give the answer as a floating-point figure rounded to two decimal places.

In [14]:
df["duration"].mean().round(2)

107.19

The average film length is 107.19 to 2 decimal places

1.5. How many films between 90 minutes and two hours long were released in 2008?

In [15]:
# Convert title_year to a date data type. It is currently a float64 type.
df["title_year"] = pd.to_datetime(df["title_year"])

In [16]:
#1.5. How many films between 90 minutes and two hours long were released in 2008?
df[((df["duration"] <= 90) & (df["duration"] >= 180)) & (df["title_year"] == "2008")]

Empty DataFrame
Columns: [color, director_name, num_critic_for_reviews, duration, director_facebook_likes, actor_3_facebook_likes, actor_2_name, actor_1_facebook_likes, gross, genres, actor_1_name, movie_title, num_voted_users, cast_total_facebook_likes, actor_3_name, facenumber_in_poster, plot_keywords, movie_imdb_link, num_user_for_reviews, language, country, content_rating, budget, title_year, actor_2_facebook_likes, imdb_score, aspect_ratio, movie_facebook_likes]
Index: []

[0 rows x 28 columns]

1.6. The `budget` column contains the film's budget. What is the median budget for all the films listed? Give the answer as an integer.

In [17]:
#budget is an object variable. This must be converted to float first
df["budget"] =df["budget"].astype(float)

In [ ]:
df["budget"] =df["budget"].astype(np.int64) # Conversion to integer

ValueError: invalid literal for int() with base 10: '237000000.0$'

In [ ]:
df["budget"] #Checking

0       237000000.0
1       300000000.0
2       245000000.0
3       250000000.0
4               0.0
           ...     
5038            0.0
5039            0.0
5040         1400.0
5041            0.0
5042         1100.0
Name: budget, Length: 5043, dtype: object

In [ ]:
df.head()

NameError: name 'df' is not defined

In [ ]:
print("The Median of budget column from the dataset = "), df['budget'].median

The Median of budget column from the dataset = 


(None,
 <bound method NDFrame._add_numeric_operations.<locals>.median of 0       237000000
 1       300000000
 2       245000000
 3       250000000
 4               0
           ...    
 5038            0
 5039            0
 5040         1400
 5041            0
 5042         1100
 Name: budget, Length: 5043, dtype: int64>)

**Task 2. Answering student questions**

How would you answer the student's question below? Your task is to get your message across in such a way that a beginner can understand your explanation. You can do this any way you want (pictures, GIFs, metaphors, anything) so long as it makes your explanation clear. Indicate how much time you spent completing this task.

> **What is the difference between DataFrame and Series?**

Series can only contain single list with index, whereas dataframe can be made of more than one series. 
A dataframe is a collection of series that can be used to analyse the data.
A pandas series is a one-dimensional data structure that comprises of key-value pair, where keys/labels are the indices and values are the values stored on that index. It is similar to a python dictionary, except it provides more freedom to manipulate and edit the data.
A pandas DataFrame is a two-dimensional data structure that can be thought of as a spreadsheet. It can also be thought of as a collection of two or more series with common indices.

image.png

**Task 3. Task on probability theory**

You are given two random variables X and Y.

E(X) = 0.5, Var(X) = 2

E(Y) = 7, Var(Y) = 3.5

cov (X, Y) = -0.8

Find the variance of the random variable Z = 2X - 3Y

**Task 4. Task for Data Science course**

Omer trained a linear regression model and tested its performance on a test sample of 500 objects. On 400 of those, the model returned a prediction higher than expected by 0.5, and on the remaining 100, the model returned a prediction lower than expected by 0.7.

What is the MSE for his model?

Limor claims that the linear regression model wasn't trained correctly, and we can do improve it by changing all the answers by a constant value. What will be her MSE?

You can assume that Limor found the smallest error under her constraints.

**Return two values - Omer's and Limor's MSE.**

Task 5. Please make a short video (under 5 minutes) and tell us more about yourself, your experience in DS/DA and the reasons you're interested in becoming a tutor with Practicum. If you're allowed to, tell us about the most interesting projects you have worked on. 